astropy/io/fits/convenience.py

```diff
@@ -601,42 +601,44 @@ def append(filename, data, header=None, checksum=False, verify=True, **kwargs):
         faster.
 
     kwargs
-        Any additional keyword arguments to be passed to
-        `astropy.io.fits.open`.
+        Additional arguments are passed to:
+        - `~astropy.io.fits.writeto` if the file does not exist or is empty. In this case
+        ``output_verify`` is the only possible argument.
+        - `~astropy.io.fits.open` if ``verify`` is True or if ``filename`` is a file object.
+        - Otherwise no additional arguments can be used.
+
     """
 
     name, closed, noexist_or_empty = _stat_filename_or_fileobj(filename)
 
-    if noexist_or_empty:
-        #
-        # The input file or file like object either doesn't exits or is
-        # empty.  Use the writeto convenience function to write the
-        # output to the empty object.
-        #
+    # The input file or file like object either doesn't exits or is
+    # empty.  Use the writeto convenience function to write the
+    # output to the empty object.
+    #
+    if noexist_or_empty and not isinstance(filename, (PrimaryHDU, HDUList)):
         writeto(filename, data, header, checksum=checksum, **kwargs)
+
     else:
         hdu = _makehdu(data, header)
 
-        if isinstance(hdu, PrimaryHDU):
+        if isinstance(hdu, (PrimaryHDU, HDUList)):
             hdu = ImageHDU(data, header)
 
-        if verify or not closed:
-            f = fitsopen(filename, mode='append')
+        if (verify or not closed) and not isinstance(filename, (PrimaryHDU, HDUList)):
+            f = fitsopen(filename, mode='append', **kwargs)
             try:
                 f.append(hdu)
-
                 # Set a flag in the HDU so that only this HDU gets a checksum
                 # when writing the file.
                 hdu._output_checksum = checksum
             finally:
                 f.close(closed=closed)
         else:
-            f = _File(filename, mode='append')
-            try:
+            f = filename
+            if kwargs:
+                warnings.warn('Unexpected **kwargs: %r' % kwargs, AstropyUserWarning)
                 hdu._output_checksum = checksum
-                hdu._writeto(f)
-            finally:
-                f.close()
+                f = hdu
```

Comment:
Why did you introduce new changes to this file ? The previous version was almost good (it was just missing the warning instead of the exception).

Reply:
Sorry, but I was testing on some file objects and ran into some problems... If I do not modify the file the file object got diverted to Case 1: `noexist_or_empty`. After I have changed by adding a conditional it ran into another issue, which is that HDU objects are not writable, which involves `f = fitsopen(filename, mode='append', **kwargs)` in Case 2, and both `f = _File(filename, mode='append')` and `hdu._writeto(f)` in Case 3.

Reply:
I would not want to bring in unnecessary changes if there is no need to...

Reply:
I think it still needs work though. For example, I will need to revert 
```
if isinstance(hdu, (PrimaryHDU, HDUList)):
    hdu = ImageHDU(data, header)
```
back to
```
if isinstance(hdu, PrimaryHDU):
    hdu = ImageHDU(data, header)
```

Reply:
@saimn Let me double check everything again... will get back to you soon.

Reply:
Think the warning snippets of the code might be problematic at this stage. Will abandon it in this PR.